bonsai-json-eval/src/main/java/com/phonepe/commons/bonsai/json/eval/JsonPathFilterEvaluationEngine.java

```diff
@@ -199,17 +211,29 @@ public Boolean visit(GenericFilter filter) {
 
 
     private <T> Boolean applyAllMatchFilter(Filter filter, TypeRef<List<T>> typeRef, Predicate<T> predicate) {
-        List<T> nonNullValues = nonNullValues(filter, typeRef);
-        return isNotEmpty(nonNullValues) && nonNullValues.stream().allMatch(predicate);
+        List<T> values = readContextValues(filter, typeRef);
+        if (!isNotEmpty(values)) {
+            return false;
+        }
+        boolean hasNonNullValue = false;
+        for (T value : values) {
+            if (value != null) {
+                hasNonNullValue = true;
+                if (!predicate.test(value)) {
+                    return false;
+                }
+            }
+        }
+        return hasNonNullValue;
     }
 
-    private <T> List<T> nonNullValues(Filter filter, TypeRef<List<T>> typeRef) {
+    private <T> List<T> readContextValues(Filter filter, TypeRef<List<T>> typeRef) {
         List<T> values = context.documentContext().read(filter.getField(), typeRef);
         if (log.isTraceEnabled()) {
             log.trace(BONSAI_FILTER_VALUES_DOCUMENT_LOG_STR, context.id(), filter.getField(), values,
                       context.documentContext().json());
         }
-        return values == null ? Collections.emptyList() : values.stream().filter(Objects::nonNull).toList();
+        return values == null ? Collections.emptyList() : values;
```

Comment:
Why not filter for non null here? This then needs all callers to check for `null`.

Reply:
This stream to list each time, came out expensive in the perf as well. The callers iterate again over this list. Hence, took the decision of moving it to the callers, each do it in their loops/streams in a more optimal way.